apps/s_server.c

```diff
@@ -3046,8 +3048,36 @@ static int www_body(int s, int stype, int prot, unsigned char *context)
         SSL_set_msg_callback_arg(con, bio_s_msg ? bio_s_msg : bio_s_out);
     }
 
+    if (early_data) {
+        int edret = SSL_READ_EARLY_DATA_ERROR;
+        size_t readbytes;
+
+        while (edret != SSL_READ_EARLY_DATA_FINISH) {
+            for (;;) {
+                edret = SSL_read_early_data(con, buf, bufsize, &readbytes);
+                if (edret != SSL_READ_EARLY_DATA_ERROR)
+                    break;
+
+                switch (SSL_get_error(con, 0)) {
+                case SSL_ERROR_WANT_WRITE:
+                case SSL_ERROR_WANT_ASYNC:
+                case SSL_ERROR_WANT_READ:
+                    /* Just keep trying - busy waiting */
+                    continue;
+                default:
+                    BIO_printf(bio_err, "Error reading early data\n");
+                    ERR_print_errors(bio_err);
+                    goto err;
+                }
+            }
+            if (readbytes > 0) {
+                BIO_write(edio, buf, (int)readbytes);
+            }
+        }
+    }
+
     for (;;) {
-        i = BIO_gets(io, buf, bufsize - 1);
```

Comment:
This is proceeding to use the early data, potentially prior to the completion of the handshake, without verifying that the request in early data is safe (in the HTTP sense) and replay-safe, which is incompatible with RFC 8470's requirements.
However, `s_server -www` is not actually a real HTTP server and offers support for only a few variants of GET and no other HTTP verbs, so IIUC we do implicitly meet the requirements of RFC 8470.  I think we should put a comment here indicating that we are compliant with RFC 8470's requirements for handling requests in early data by virtue of `s_server` only ever handling requests that are safe to process.

Reply:
Would it be possible to also go a step further and have the early data buffered and processed after the handshake?

Reply:
I almost suggested that, actually -- from my first inspection it looked like just a call to `SSL_do_handshake()` between loops would do the trick, though I didn't actually confirm that.
Waiting until the handshake is complete before processing early data is always safe, regardless of the type of data conveyed.

Reply:
> replay-safe

OpenSSL's early data implementation automatically operates in replay safe mode by default. You have to explicitly switch it off with `SSL_OP_NO_ANTI_REPLAY` (or using the `-no_anti_replay` s_server option).

Reply:
The TLS layer's replay-safe operation is orthogonal to the HTTP-layer RFC 8470 compliance.  Merely using a replay cache at the TLS layer does not make it permitted to process arbitrary HTTP requests in early data.